core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
```

Comment:
`mapping: Map[Int, Int]` is technically an implementation detail, and as such maybe should not be exposed as `val`. Perhaps `Perm` should be a trait?

Reply:
Good point. Originally, this was a case class with an exposed `map` element.
I could make this `private[Perm]` and move the `Group.combine` implementation that depends on it.